src/main/java/net/dv8tion/jda/api/requests/Route.java

```diff
@@ -329,100 +315,317 @@ public static Route custom(@Nonnull Method method, @Nonnull String route)
+     *
+     * @return The route template
+     */
+    @Nonnull
     public String getRoute()
     {
-        return route;
+        return String.join("/", template);
     }
 
+    /**
+     * The number of parameters for this route, not including query parameters.
+     *
+     * @return The parameter count
+     */
     public int getParamCount()
     {
         return paramCount;
     }
 
-    public CompiledRoute compile(String... params)
+    /**
+     * Compile the route with provided parameters.
+     * <br>The number of parameters must match the number of placeholders in the route template.
+     * The provided arguments are positional and will replace the placeholders of the template in order of appearance.
+     *
+     * <p>Use {@link CompiledRoute#withQueryParams(String...)} to add query parameters to the route.
+     *
+     * @param  params
+     *         The parameters to compile the route with
+     *
+     * @throws IllegalArgumentException
+     *         If the number of parameters does not match the number of placeholders, or null is provided
+     *
+     * @return The compiled route, ready to use for rate-limit handling
+     */
+    @Nonnull
+    public CompiledRoute compile(@Nonnull String... params)
     {
-        if (params.length != paramCount)
-        {
-            throw new IllegalArgumentException("Error Compiling Route: [" + route + "], incorrect amount of parameters provided." +
-                    "Expected: " + paramCount + ", Provided: " + params.length);
-        }
-
-        //Compile the route for interfacing with discord.
-        Set<String> major = new HashSet<>();
-        StringBuilder compiledRoute = new StringBuilder(route);
-        for (int i = 0; i < paramCount; i++)
+        Checks.noneNull(params, "Arguments");
+        Checks.check(
+            params.length == paramCount,
+            "Error Compiling Route: [%s], incorrect amount of parameters provided. Expected: %d, Provided: %d",
+            this, paramCount, params.length
+        );
+
+        StringJoiner major = new StringJoiner(":").setEmptyValue("n/a");
+        StringJoiner compiledRoute = new StringJoiner("/");
+
+        int paramIndex = 0;
+        for (String element : template)
         {
-            int paramStart = compiledRoute.indexOf("{");
-            int paramEnd = compiledRoute.indexOf("}");
-            String paramName = compiledRoute.substring(paramStart+1, paramEnd);
-            if (majorParameters.contains(paramName))
+            if (element.charAt(0) == '{')
             {
-                if (params[i].length() > 30) // probably a long interaction_token, hash it to keep logs clean (not useful anyway)
-                    major.add(paramName + "=" + Integer.toUnsignedString(params[i].hashCode()));
-                else
-                    major.add(paramName + "=" + params[i]);
+                String name = element.substring(1, element.length() - 1);
```

Comment:
Can we do a check somewhere, like an assert, to ensure there _is_ a closing `}`?

Reply:
Since the routes system is getting pulled into API, the Custom routes are more likely to get touched, so checking for mismatched `{`/`}` would be nice

Reply:
This is already checked in the route constructor.

```
Checks.check(paramCount == Helpers.countMatches(route, '}'), "An argument does not have both {}'s for route: %s %s", method, route);
```

Reply:
This is checked in the route constructor already, no need to check it on every compile